{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import re\n",
    "import json\n",
    "import pickle\n",
    "import datetime\n",
    "import time\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.common.exceptions import TimeoutException\n",
    "from selenium.webdriver.firefox.firefox_binary import FirefoxBinary\n",
    "from selenium.webdriver.common.desired_capabilities import DesiredCapabilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "binary = FirefoxBinary(r'/usr/bin/firefox')\n",
    "caps = DesiredCapabilities.FIREFOX.copy()\n",
    "#Set ‘marionette’ browser to True\n",
    "caps['marionette'] = True\n",
    "#Launch the Firefox instance by specifying the geckodriver executable path\n",
    "#driver = webdriver.Firefox(firefox_binary=binary,capabilities=caps)\n",
    "#driver.wait = WebDriverWait(driver, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formatOriginalNameToWikiName(originalname):\n",
    "    \"\"\"\n",
    "    return goodname if we return a better format from wiki\n",
    "    if not we just return empty string\n",
    "    we need this function,since if not formated yet, most of the time when you search wiki with the bad \n",
    "    name in the url it will return nothing.\n",
    "    \"\"\"\n",
    "    wikipedia.set_lang(\"en\")\n",
    "    allWikiResults=wikipedia.search(originalname)\n",
    "    if(len(allWikiResults)==0):\n",
    "        return \"\"\n",
    "    else:\n",
    "        return str(wikipedia.search(originalname)[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#inorder to be parallel we need to pass in a new driver\n",
    "def MakeSeleniumToSearchWithOriginalName(originalname,driver):\n",
    "    wikiname=formatOriginalNameToWikiName(originalname)\n",
    "    nametosearch=originalname if(wikiname==\"\") else wikiname\n",
    "    driver.get(\"https://en.wikipedia.org/wiki/\"+str(nametosearch))\n",
    "    result={}\n",
    "    result[\"findresult\"]={}\n",
    "    result[\"nofind\"]={}\n",
    "    try:\n",
    "        elem = driver.find_element_by_css_selector(\".interwiki-ar a\")\n",
    "        \"\"\"\n",
    "        these two lines of code needs to run before elem.click(), since it will goto\n",
    "        another page never find it any more.\n",
    "        \"\"\"\n",
    "        tempdic={}\n",
    "        #print(\"me\"+str(elem.get_attribute(\"href\")))\n",
    "        tempdic[\"arurl\"]=str(elem.get_attribute(\"href\"))\n",
    "        elem.click()\n",
    "        tempdic[\"originalname\"]=originalname\n",
    "        tempdic[\"wikiname\"]=wikiname\n",
    "        firstheading=driver.find_element_by_id(\"firstHeading\")\n",
    "        #arabic is from left to right that why u need to get the first one that returns.\n",
    "        tempdic[\"arname\"]=firstheading.text.split(\"\\n\")[0]\n",
    "        #print(tempdic[\"arname\"])\n",
    "        result[\"findresult\"]=tempdic\n",
    "    except Exception as e:\n",
    "        #print(e)\n",
    "        tempno={}\n",
    "        tempno[\"originalname\"]=originalname\n",
    "        result[\"nofind\"]=tempno\n",
    "        pass\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def clean_line(line):\n",
    "    # Take out extra space, underscores, comments, etc.\n",
    "    cleaned = re.sub(\"_* .+\", \"\", line).strip()\n",
    "    cleaned = re.sub(\"_$\", \"\", cleaned, flags=re.MULTILINE)\n",
    "    return cleaned\n",
    "\n",
    "def ingest_dictionary(dict_path):\n",
    "    \"\"\"\n",
    "    Read in the country (or other) actor dictionaries.\n",
    "    \"\"\"\n",
    "    with open(dict_path) as f:\n",
    "        country_file = f.read()\n",
    "    split_file = country_file.split(\"\\n\")\n",
    "    \n",
    "    dict_dict = []\n",
    "    key_name = \"\"\n",
    "    alt_names = [] \n",
    "    roles = []\n",
    "\n",
    "    for line in split_file:\n",
    "        if not line:\n",
    "            pass\n",
    "        elif line[0] == \"#\":\n",
    "            pass\n",
    "        elif re.match(\"[A-Z]\", line[0]):\n",
    "            # handle the previous\n",
    "            entry = {\"actor_en\" : key_name,\n",
    "                    \"alt_names_en\" : alt_names,\n",
    "                    \"roles\" : roles}\n",
    "            dict_dict.append(entry)\n",
    "            # zero everything out\n",
    "            alt_names = []\n",
    "            roles = []\n",
    "            # make new key name\n",
    "            key_name = clean_line(line)\n",
    "            # check to see if the role is built in\n",
    "            if bool(re.search(\"\\[[A-Z]{3}\\]\", line)):\n",
    "                roles = re.findall(\"\\[(.+?)\\]\", line)\n",
    "        elif line[0] == \"+\":\n",
    "            cleaned = clean_line(line[1:])\n",
    "            alt_names.append(cleaned)\n",
    "        elif re.match(\"\\s\", line):\n",
    "            roles.append(line.strip())\n",
    "    return dict_dict \n",
    "dp = \"./Phoenix.Countries.actors.txt\"\n",
    "dict_dict = ingest_dictionary(dp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18390"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dict_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [],
   "source": [
    "def buildMultiLanguageActorDictionary(dict_dict):\n",
    "    finalResult={}\n",
    "    finalResult[\"goodones\"]=[]\n",
    "    finalResult[\"badones\"]=[]\n",
    "    driver = webdriver.Firefox(firefox_binary=binary,capabilities=caps)\n",
    "    driver.wait = WebDriverWait(driver, 5)\n",
    "    for item in dict_dict:\n",
    "        originalname=item[\"actor_en\"]\n",
    "        if(originalname!=\"\"):\n",
    "            temp=MakeSeleniumToSearchWithOriginalName(originalname,driver)\n",
    "            if(temp[\"findresult\"]):\n",
    "                finalResult[\"goodones\"].append(temp[\"findresult\"])\n",
    "            else:\n",
    "                finalResult[\"badones\"].append(temp[\"nofind\"])\n",
    "    return finalResult\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18390"
      ]
     },
     "execution_count": 310,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dict_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "driver 1 start\n",
      "driver 2 start\n",
      "driver 3 start\n",
      "driver 4 start\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process Process-7:\n",
      "Process Process-9:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 141, in _new_conn\n",
      "    (self.host, self.port), self.timeout, **extra_kw)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/util/connection.py\", line 60, in create_connection\n",
      "    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):\n",
      "  File \"/usr/lib/python3.5/socket.py\", line 732, in getaddrinfo\n",
      "    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 141, in _new_conn\n",
      "    (self.host, self.port), self.timeout, **extra_kw)\n",
      "socket.gaierror: [Errno -3] Temporary failure in name resolution\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/util/connection.py\", line 60, in create_connection\n",
      "    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):\n",
      "  File \"/usr/lib/python3.5/socket.py\", line 732, in getaddrinfo\n",
      "    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "socket.gaierror: [Errno -3] Temporary failure in name resolution\n",
      "Traceback (most recent call last):\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 600, in urlopen\n",
      "    chunked=chunked)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 356, in _make_request\n",
      "    conn.request(method, url, **httplib_request_kw)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 600, in urlopen\n",
      "    chunked=chunked)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1106, in request\n",
      "    self._send_request(method, url, body, headers)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1151, in _send_request\n",
      "    self.endheaders(body)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 356, in _make_request\n",
      "    conn.request(method, url, **httplib_request_kw)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1102, in endheaders\n",
      "    self._send_output(message_body)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1106, in request\n",
      "    self._send_request(method, url, body, headers)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 934, in _send_output\n",
      "    self.send(msg)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1151, in _send_request\n",
      "    self.endheaders(body)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 877, in send\n",
      "    self.connect()\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1102, in endheaders\n",
      "    self._send_output(message_body)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 934, in _send_output\n",
      "    self.send(msg)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 877, in send\n",
      "    self.connect()\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 166, in connect\n",
      "    conn = self._new_conn()\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 166, in connect\n",
      "    conn = self._new_conn()\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 150, in _new_conn\n",
      "    self, \"Failed to establish a new connection: %s\" % e)\n",
      "urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7ff21fac1898>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/adapters.py\", line 440, in send\n",
      "    timeout=timeout\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 649, in urlopen\n",
      "    _stacktrace=sys.exc_info()[2])\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/util/retry.py\", line 388, in increment\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='en.wikipedia.org', port=80): Max retries exceeded with url: /w/api.php?action=query&srlimit=10&limit=10&srprop=&format=json&list=search&srsearch=GIOVANNI_SPADOLINI (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff21fac1898>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 150, in _new_conn\n",
      "    self, \"Failed to establish a new connection: %s\" % e)\n",
      "urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7ff21fab1278>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/adapters.py\", line 440, in send\n",
      "    timeout=timeout\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 649, in urlopen\n",
      "    _stacktrace=sys.exc_info()[2])\n",
      "  File \"<ipython-input-318-da492f6c002b>\", line 16, in func3\n",
      "    rst3=buildMultiLanguageActorDictionary(dict_dict[8000:12000])\n",
      "  File \"<ipython-input-315-5b46402ef819>\", line 10, in buildMultiLanguageActorDictionary\n",
      "    temp=MakeSeleniumToSearchWithOriginalName(originalname,driver)\n",
      "  File \"<ipython-input-317-df5ad56cea67>\", line 3, in MakeSeleniumToSearchWithOriginalName\n",
      "    wikiname=formatOriginalNameToWikiName(originalname)\n",
      "  File \"<ipython-input-242-916760ec7eb8>\", line 9, in formatOriginalNameToWikiName\n",
      "    allWikiResults=wikipedia.search(originalname)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/util.py\", line 28, in __call__\n",
      "    ret = self._cache[key] = self.fn(*args, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/wikipedia.py\", line 103, in search\n",
      "    raw_results = _wiki_request(search_params)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/wikipedia.py\", line 737, in _wiki_request\n",
      "    r = requests.get(API_URL, params=params, headers=headers)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/api.py\", line 72, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/util/retry.py\", line 388, in increment\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/api.py\", line 58, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='en.wikipedia.org', port=80): Max retries exceeded with url: /w/api.php?action=query&srlimit=10&limit=10&srprop=&format=json&list=search&srsearch=LIAMINE_Z%1AROUAL (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff21fab1278>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/sessions.py\", line 502, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"<ipython-input-318-da492f6c002b>\", line 5, in func1\n",
      "    rst1=buildMultiLanguageActorDictionary(dict_dict[0:4000])\n",
      "  File \"<ipython-input-315-5b46402ef819>\", line 10, in buildMultiLanguageActorDictionary\n",
      "    temp=MakeSeleniumToSearchWithOriginalName(originalname,driver)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  File \"<ipython-input-317-df5ad56cea67>\", line 3, in MakeSeleniumToSearchWithOriginalName\n",
      "    wikiname=formatOriginalNameToWikiName(originalname)\n",
      "  File \"<ipython-input-242-916760ec7eb8>\", line 9, in formatOriginalNameToWikiName\n",
      "    allWikiResults=wikipedia.search(originalname)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/util.py\", line 28, in __call__\n",
      "    ret = self._cache[key] = self.fn(*args, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/wikipedia.py\", line 103, in search\n",
      "    raw_results = _wiki_request(search_params)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/sessions.py\", line 612, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/wikipedia.py\", line 737, in _wiki_request\n",
      "    r = requests.get(API_URL, params=params, headers=headers)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/api.py\", line 72, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/adapters.py\", line 504, in send\n",
      "    raise ConnectionError(e, request=request)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/api.py\", line 58, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/sessions.py\", line 502, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "requests.exceptions.ConnectionError: HTTPConnectionPool(host='en.wikipedia.org', port=80): Max retries exceeded with url: /w/api.php?action=query&srlimit=10&limit=10&srprop=&format=json&list=search&srsearch=GIOVANNI_SPADOLINI (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff21fac1898>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/sessions.py\", line 612, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/adapters.py\", line 504, in send\n",
      "    raise ConnectionError(e, request=request)\n",
      "requests.exceptions.ConnectionError: HTTPConnectionPool(host='en.wikipedia.org', port=80): Max retries exceeded with url: /w/api.php?action=query&srlimit=10&limit=10&srprop=&format=json&list=search&srsearch=LIAMINE_Z%1AROUAL (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff21fab1278>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))\n",
      "Process Process-10:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 141, in _new_conn\n",
      "    (self.host, self.port), self.timeout, **extra_kw)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/util/connection.py\", line 60, in create_connection\n",
      "    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):\n",
      "  File \"/usr/lib/python3.5/socket.py\", line 732, in getaddrinfo\n",
      "    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):\n",
      "socket.gaierror: [Errno -3] Temporary failure in name resolution\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 600, in urlopen\n",
      "    chunked=chunked)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 356, in _make_request\n",
      "    conn.request(method, url, **httplib_request_kw)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1106, in request\n",
      "    self._send_request(method, url, body, headers)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1151, in _send_request\n",
      "    self.endheaders(body)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1102, in endheaders\n",
      "    self._send_output(message_body)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 934, in _send_output\n",
      "    self.send(msg)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 877, in send\n",
      "    self.connect()\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 166, in connect\n",
      "    conn = self._new_conn()\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 150, in _new_conn\n",
      "    self, \"Failed to establish a new connection: %s\" % e)\n",
      "urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7ff21fac11d0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/adapters.py\", line 440, in send\n",
      "    timeout=timeout\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 649, in urlopen\n",
      "    _stacktrace=sys.exc_info()[2])\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/util/retry.py\", line 388, in increment\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='en.wikipedia.org', port=80): Max retries exceeded with url: /w/api.php?action=query&srlimit=10&limit=10&srprop=&format=json&list=search&srsearch=DAVID_OSBORNE_HAY (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff21fac11d0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"<ipython-input-318-da492f6c002b>\", line 21, in func4\n",
      "    rst4=buildMultiLanguageActorDictionary(dict_dict[12000:len(dict_dict)])\n",
      "  File \"<ipython-input-315-5b46402ef819>\", line 10, in buildMultiLanguageActorDictionary\n",
      "    temp=MakeSeleniumToSearchWithOriginalName(originalname,driver)\n",
      "  File \"<ipython-input-317-df5ad56cea67>\", line 3, in MakeSeleniumToSearchWithOriginalName\n",
      "    wikiname=formatOriginalNameToWikiName(originalname)\n",
      "  File \"<ipython-input-242-916760ec7eb8>\", line 9, in formatOriginalNameToWikiName\n",
      "    allWikiResults=wikipedia.search(originalname)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/util.py\", line 28, in __call__\n",
      "    ret = self._cache[key] = self.fn(*args, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/wikipedia.py\", line 103, in search\n",
      "    raw_results = _wiki_request(search_params)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/wikipedia.py\", line 737, in _wiki_request\n",
      "    r = requests.get(API_URL, params=params, headers=headers)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/api.py\", line 72, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/api.py\", line 58, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/sessions.py\", line 502, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/sessions.py\", line 612, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/adapters.py\", line 504, in send\n",
      "    raise ConnectionError(e, request=request)\n",
      "requests.exceptions.ConnectionError: HTTPConnectionPool(host='en.wikipedia.org', port=80): Max retries exceeded with url: /w/api.php?action=query&srlimit=10&limit=10&srprop=&format=json&list=search&srsearch=DAVID_OSBORNE_HAY (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff21fac11d0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))\n",
      "Process Process-8:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 141, in _new_conn\n",
      "    (self.host, self.port), self.timeout, **extra_kw)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/util/connection.py\", line 60, in create_connection\n",
      "    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):\n",
      "  File \"/usr/lib/python3.5/socket.py\", line 732, in getaddrinfo\n",
      "    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):\n",
      "socket.gaierror: [Errno -3] Temporary failure in name resolution\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 600, in urlopen\n",
      "    chunked=chunked)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 356, in _make_request\n",
      "    conn.request(method, url, **httplib_request_kw)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1106, in request\n",
      "    self._send_request(method, url, body, headers)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1151, in _send_request\n",
      "    self.endheaders(body)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 1102, in endheaders\n",
      "    self._send_output(message_body)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 934, in _send_output\n",
      "    self.send(msg)\n",
      "  File \"/usr/lib/python3.5/http/client.py\", line 877, in send\n",
      "    self.connect()\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 166, in connect\n",
      "    conn = self._new_conn()\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connection.py\", line 150, in _new_conn\n",
      "    self, \"Failed to establish a new connection: %s\" % e)\n",
      "urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7ff21f9c9e48>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/adapters.py\", line 440, in send\n",
      "    timeout=timeout\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/connectionpool.py\", line 649, in urlopen\n",
      "    _stacktrace=sys.exc_info()[2])\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/urllib3/util/retry.py\", line 388, in increment\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='en.wikipedia.org', port=80): Max retries exceeded with url: /w/api.php?action=query&srlimit=10&limit=10&srprop=&format=json&list=search&srsearch=GUILLERMO_RODRIGUEZ_LARA (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff21f9c9e48>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"<ipython-input-318-da492f6c002b>\", line 11, in func2\n",
      "    rst2=buildMultiLanguageActorDictionary(dict_dict[4000:8000])\n",
      "  File \"<ipython-input-315-5b46402ef819>\", line 10, in buildMultiLanguageActorDictionary\n",
      "    temp=MakeSeleniumToSearchWithOriginalName(originalname,driver)\n",
      "  File \"<ipython-input-317-df5ad56cea67>\", line 3, in MakeSeleniumToSearchWithOriginalName\n",
      "    wikiname=formatOriginalNameToWikiName(originalname)\n",
      "  File \"<ipython-input-242-916760ec7eb8>\", line 9, in formatOriginalNameToWikiName\n",
      "    allWikiResults=wikipedia.search(originalname)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/util.py\", line 28, in __call__\n",
      "    ret = self._cache[key] = self.fn(*args, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/wikipedia.py\", line 103, in search\n",
      "    raw_results = _wiki_request(search_params)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/wikipedia/wikipedia.py\", line 737, in _wiki_request\n",
      "    r = requests.get(API_URL, params=params, headers=headers)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/api.py\", line 72, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/api.py\", line 58, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/sessions.py\", line 502, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/sessions.py\", line 612, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/home/yan/python_vir/env/lib/python3.5/site-packages/requests/adapters.py\", line 504, in send\n",
      "    raise ConnectionError(e, request=request)\n",
      "requests.exceptions.ConnectionError: HTTPConnectionPool(host='en.wikipedia.org', port=80): Max retries exceeded with url: /w/api.php?action=query&srlimit=10&limit=10&srprop=&format=json&list=search&srsearch=GUILLERMO_RODRIGUEZ_LARA (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff21f9c9e48>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))\n"
     ]
    }
   ],
   "source": [
    "from multiprocessing import Process\n",
    "\n",
    "def func1():\n",
    "      print('driver 1 start')\n",
    "      rst1=buildMultiLanguageActorDictionary(dict_dict[0:40])\n",
    "      wholeresults.append(rst1)\n",
    "      print('driver 1 finish')\n",
    "\n",
    "def func2():\n",
    "      print('driver 2 start')\n",
    "      rst2=buildMultiLanguageActorDictionary(dict_dict[40:80])\n",
    "      wholeresults.append(rst2)\n",
    "      print('driver 2 finish')\n",
    "def func3():\n",
    "      print('driver 3 start')\n",
    "      rst3=buildMultiLanguageActorDictionary(dict_dict[80:120])\n",
    "      wholeresults.append(rst3)\n",
    "      print('driver 3 finish')\n",
    "def func4():\n",
    "      print('driver 4 start')\n",
    "      rst4=buildMultiLanguageActorDictionary(dict_dict[120:160])#len(dict_dict)])\n",
    "      wholeresults.append(rst4)\n",
    "      print('driver 4 finish')\n",
    "        \n",
    "\n",
    "if __name__ == '__main__':\n",
    "      wholeresults=[]\n",
    "      p1 = Process(target=func1)\n",
    "      p1.start()\n",
    "      p2 = Process(target=func2)\n",
    "      p2.start()\n",
    "      p3= Process(target=func3)\n",
    "      p3.start()\n",
    "      p4 = Process(target=func4)\n",
    "      p4.start()\n",
    "      \n",
    "      p1.join()\n",
    "      p2.join()\n",
    "      p3.join()\n",
    "      p4.join()\n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "test=len(wikipedia.search(\"obama\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Barack Obama'"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wikipedia.search(\"obama\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "MakeSeleniumToSearchWithWikiFormattedName(\"Mohammad Najibullah\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "driver.get(\"https://en.wikipedia.org/wiki/\"+\"Mohammad Najibullah\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "elem = driver.find_element_by_css_selector(\".interwiki-ar a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://ar.wikipedia.org/wiki/%D9%85%D8%AD%D9%85%D8%AF_%D9%86%D8%AC%D9%8A%D8%A8_%D8%A7%D9%84%D9%84%D9%87'"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "elem.get_attribute(\"href\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "test=driver.find_element_by_id(\"firstHeading\")\n",
    "#arabic is from left to right that why u need to get the first one that returns.\n",
    "test.text.split(\"\\n\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'محمد نجيب الله'"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#arabic is from left to right that why u need to get the first one that returns.\n",
    "test.text.split(\"\\n\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'https://en.wikipedia.org/wiki/Sibghatullah_Mojaddedi'\n"
     ]
    }
   ],
   "source": [
    "from googleapiclient.discovery import build\n",
    "import pprint\n",
    "\n",
    "my_api_key = \"AIzaSyBBulleVoiDN9i8NITQqH_BUNGgyWX-nmA\"\n",
    "my_cse_id = \"003461024781403571159:p4qrcenq1l0\"\n",
    "\n",
    "def google_search(search_term, api_key, cse_id, **kwargs):\n",
    "    service = build(\"customsearch\", \"v1\", developerKey=api_key)\n",
    "    res = service.cse().list(q=search_term, cx=cse_id, **kwargs).execute()\n",
    "    #return res['spelling']['correctedQuery']\n",
    "#     return res['item']\n",
    "    return res['items']\n",
    "results = google_search(\n",
    "    'SIBGHATULLAH_MOJADEDI', my_api_key, my_cse_id, num=1)\n",
    "for result in results:\n",
    "     pprint.pprint(result['formattedUrl'])\n",
    "#print(results)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ألبرت أينشتاين\n"
     ]
    }
   ],
   "source": [
    "print(\"\\u0623\\u0644\\u0628\\u0631\\u062a \\u0623\\u064a\\u0646\\u0634\\u062a\\u0627\\u064a\\u0646\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "payload = {'action': 'query', 'titles': 'Alert Einstein','prop':'langlinks','format':'json'}\n",
    "\n",
    "r = requests.get(\"https://en.wikipedia.org/w/api.php\", data=payload)\n",
    "soup=BeautifulSoup(r.content,\"lxml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wikipedia ( or  WIK-i-PEE-dee-ə) is a free online encyclopedia with the aim to allow anyone to edit articles. Wikipedia is the largest and most popular general reference work on the Internet and is ranked among the ten most popular websites. Wikipedia is owned by the nonprofit Wikimedia Foundation.\n",
      "Wikipedia was launched on January 15, 2001, by Jimmy Wales and Larry Sanger. Sanger coined its name, a portmanteau of wiki and encyclopedia. There was only the English language version initially, but it quickly developed similar versions in other languages, which differ in content and in editing practices. With 5,433,361 articles, the English Wikipedia is the largest of the more than 290 Wikipedia encyclopedias. Overall, Wikipedia consists of more than 40 million articles in more than 250 different languages and, as of February 2014, it had 18 billion page views and nearly 500 million unique visitors each month.\n",
      "As of March 2017, Wikipedia has about forty thousand high-quality articles known as Featured Articles and Good Articles that cover vital topics. In 2005, Nature published a peer review comparing 42 science articles from Encyclopædia Britannica and Wikipedia, and found that Wikipedia's level of accuracy approached Encyclopædia Britannica's.\n",
      "Wikipedia has been criticized for allegedly exhibiting systemic bias, presenting a mixture of \"truths, half truths, and some falsehoods\", and, in controversial topics, being subject to manipulation and spin.\n"
     ]
    }
   ],
   "source": [
    "import wikipedia\n",
    "print(wikipedia.summary(\"Wikipedia\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Barack Obama',\n",
       " 'Barack Obama in comics',\n",
       " 'Barack Obama Sr.',\n",
       " 'Barack Obama: Der schwarze Kennedy',\n",
       " 'List of things named after Barack Obama',\n",
       " 'Inauguration of Barack Obama',\n",
       " 'Bibliography of Barack Obama',\n",
       " 'Barack Obama Presidential Center',\n",
       " 'Timeline of the presidency of Barack Obama',\n",
       " 'Barack Obama religion conspiracy theories']"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wikipedia.search(\"Barack\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "wikipedia.set_lang(\"en\")\n",
    "test=wikipedia.page(\"SIBGHATULLAH_MOJADEDI\").html()\n",
    "soup=BeautifulSoup(test,'lxml')\n",
    "hi=soup.find(\"li\",{\"class\":\"interwiki-ar\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['http://www.worldcat.org/identities/containsVIAFID/75918762',\n",
       " 'http://www.worldcat.org/oclc/123336516',\n",
       " 'http://www.worldcat.org/oclc/237144347',\n",
       " 'http://aviation-safety.net/database/record.php?id=19920529-0',\n",
       " 'http://hrw.org/reports/2005/afghanistan0605/4.htm#_Toc105552342',\n",
       " 'http://id.loc.gov/authorities/names/no97021045',\n",
       " 'http://www.afghan-bios.info/index.php?option=com_afghanbios&id=1085&task=view&total=2314&start=1266&Itemid=2',\n",
       " 'http://www.aftabir.com/news/article/view/2016/02/09/1139108',\n",
       " 'http://www.bbc.com/pashto/afghanistan/2016/02/160215_hh-27th-anniv-soviet-forces-defeat-afg',\n",
       " 'http://www.khaama.com/mojadedi-announces-the-establishment-of-a-new-political-council-9607',\n",
       " 'http://www.mojaddedi.org/biography-of-sibghatullah-al-mojaddedi.html',\n",
       " 'http://www.pts.af/',\n",
       " 'http://www.rferl.org/content/article/1066619.html',\n",
       " 'http://www.washingtontimes.com/news/2010/sep/28/afghan-peace-council-draws-fire/',\n",
       " 'http://www.zmong-afghanistan.com/profiles/sibghatullah.asp',\n",
       " 'https://archive.org/stream/azu_acku_risalah_ds371_2_meem46_yaa1375#page/n1/mode/1up',\n",
       " 'https://books.google.com.my/books?id=1xyh_DBV1bMC&pg=PA492&lpg=PA492&dq=sibghatullah+mujaddidi+born&source=bl&ots=0-bbq_LRo5&sig=evfzzrgRMTkeWS13W4QhfaHJwe4&hl=en&sa=X&redir_esc=y#v=onepage&q=sibghatullah%20mujaddidi%20born&f=false',\n",
       " 'https://books.google.com/books?id=RUSNyMH1aFQC&lpg=PR4&pg=PA406#v=onepage&q&f=false',\n",
       " 'https://books.google.com/books?id=_zWhhy8L0uQC&lpg=PP1&pg=PT15#v=onepage&q&f=false',\n",
       " 'https://viaf.org/viaf/75918762',\n",
       " 'https://web.archive.org/web/20110606152711/http://www.zmong-afghanistan.com/profiles/sibghatullah.asp']"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wikipedia.page(\"SIBGHATULLAH_MOJADEDI\").references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Sibghatullah Mojaddedi (Pashto: صبغت الله مجددی\\u200e\\u200e, born 21 April 1925) is a politician in Afghanistan, who served as Acting President after the fall of Mohammad Najibullah's government in April 1992. He is also the founder of the Afghan National Liberation Front, and served as the chairman of the 2003 loya jirga that approved Afghanistan's new constitution.\""
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ny = wikipedia.page(\"New York\")\n",
    "#ny.title\n",
    "#ny.url\n",
    "#ny.links[0]\n",
    "#wikipedia.set_lang(\"en\")\n",
    "wikipedia.summary(\"SIBGHATULLAH_MOJADEDI\", sentences=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Mohammad Najibullah',\n",
       " 'Mohammad Najatuallah Siddiqui',\n",
       " 'Abdul Razzaq (Taliban governor)',\n",
       " 'Vice President of Afghanistan',\n",
       " 'Abdul Wahed Sorabi',\n",
       " 'Habibia High School',\n",
       " 'Ghazi High School',\n",
       " 'Najib',\n",
       " 'Najibullah Torwayana',\n",
       " 'National Reconciliation']"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wikipedia.search(\"MOHAMMAD_NAJIBULLAH\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
